rfcs/20201121-keras-model-fit-ps.md

```diff
@@ -368,11 +368,54 @@ Since the workers will sync every epoch anyway, fetching the remote values incur
+
+**Option 2: A `step` is one batch on every worker**
+
+With this mental model, every time `Model.train_function` is called, it schedules one batch to execute on each worker. This means that if there are `W` workers, passing `steps_per_epoch=100` will actually run `100 * W` batches of training, with each worker seeing `100` batches.
+
+**Pros:**
+
+- Batch-level `Callback`s can be supported with this approach. We can sync every step, and we can rely on the `steps_per_execution` parameter to ensure that the step time is large enough so that this syncing is not prohibitively expensive.
+
+**Cons:**
+
+- Users need to be aware of how many workers are in their cluster when setting `steps_per_epoch`. Setting `steps_per_epoch=100` will actually run `100 * W` training batches, which could be confusing to users. Halving the number of workers, without changing any code, would also halve the amount of training performed.
+
+If we go with Option (1), we should disallow batch-level `Callback`s, since in this case `ParameterServerStrategy` with batch-level `Callback`s will always be slower than training on a single machine.
+
+If we go with Option (2) we should support batch-level `Callback`s, but we will use existing logic in `CallbackList` to detect when batch-level `Callback`s are passed, and only incur the performance penalty of syncing workers each batch when the user has passed batch-level `Callback`s (for context, none of Keras's built-in `Callbacks` other than the `ProgressBar` will incur this penalty). This logic was originally put in place to ensure that TPU async mode was only blocked when needed, and applies equally well to `ParameterServerStrategy` without significant modifications. We will also re-use existing logic to log a warning to the user when their batch-level `Callback`s are causing a significant slowdown in training time. This logic also resides in the `CallbackList` object.
+
+**Asynchronous Callbacks**
+
+If we wanted to support batch-level `Callback`s with the `step` mental model Option 1 above, we could alternatively introduce a new concept to `Model.fit` of "asynchronous callbacks". An asynchronous `Callback` wouldn't block until its results were ready, instead we would schedule all batches upfront, and then send the results of the batches to the asynchronous `Callback` when the `Coordinator` determines that the batch has finished executing.
+
+**Pros:**
+
+- This allows for a form of batch-level `Callback`s that doesn't need to block. It would work with either mental model of `steps` outlined above. It would work well for certain tasks such as `Checkpointing`.
+
+**Cons:**
+
+- Asynchronous `Callback`s would be limited in what they could do. Any changes that an asynchronous object makes to a `tf.Variable` (such as the `learning_rate`) would not take effect until the next epoch, since all of the batches were already scheduled before the `Callback` executes.
```

Comment:
I think some of the scheduled functions in the same epoch can still pick up the updated variables because the variable read happens at function execution, but not at scheduling time (unless scheduled functions execute super quickly and callback's variable updates are super slow).